worlds/sc2/rules.py

```diff
@@ -1331,6 +1392,61 @@ def protoss_defense_rating(self, state: CollectionState, zerg_enemy: bool) -> in
 
     def protoss_common_unit(self, state: CollectionState) -> bool:
         return state.has_any(self.basic_protoss_units, self.player)
+    
+    def protoss_any_anti_air_unit_or_soa_any_protoss(self, state: CollectionState) -> bool:
+        return (
+            self.protoss_any_anti_air_unit(state)
+            or (
+                self.spear_of_adun_presence in (SpearOfAdunPresence.option_everywhere, SpearOfAdunPresence.option_protoss)
+                and self.protoss_any_anti_air_soa(state)
+            )
+        )
+    
+    def protoss_any_anti_air_unit_or_soa(self, state: CollectionState) -> bool:
+        return self.protoss_any_anti_air_unit(state) or self.protoss_any_anti_air_soa(state)
+
+    def protoss_any_anti_air_soa(self, state: CollectionState) -> bool:
+        return (
+            state.has_any((
+                item_names.SOA_ORBITAL_STRIKE, item_names.SOA_SOLAR_LANCE,
+                item_names.SOA_SOLAR_BOMBARDMENT, item_names.SOA_PURIFIER_BEAM,
+                item_names.SOA_PYLON_OVERCHARGE,
+            ), self.player)
+            or state.has(item_names.SOA_PROGRESSIVE_PROXY_PYLON, self.player, 2)  # Reinforcements)
```

Comment:
Use `state.count` >= 2 as `ValidInventory` in `pool_filter.py` can't process it correctly